sdk/eventhub/azure-eventhubs/azure/eventhub/aio/receiver_async.py

```diff
@@ -156,18 +165,46 @@ def __aiter__(self):
             self._handler = ReceiveClientAsync(
                 source,
                 auth=self.client.get_auth(**alt_creds),
-                debug=self.client.debug,
+                debug=self.client.config.network_tracing,
                 prefetch=self.prefetch,
                 link_properties=self.properties,
                 timeout=self.timeout,
                 error_policy=self.retry_policy,
                 keep_alive_interval=self.keep_alive,
                 client_name=self.name,
-                properties=self.client.create_properties(),
+                properties=self.client.create_properties(self.client.config.user_agent),
                 loop=self.loop)
-        await self._handler.open_async()
-        while not await self._handler.client_ready_async():
-            await asyncio.sleep(0.05)
+
+        if not self.running:
+            try:
+                await self._handler.open_async()
+                self.running = True
+                while not await self._handler.client_ready_async():
+                    await asyncio.sleep(0.05)
+            except errors.AuthenticationException:
+                    log.info("Receiver failed authentication. Retrying...")
+                    await self.reconnect()
+            except (errors.LinkDetach, errors.ConnectionClose) as shutdown:
+                if shutdown.action.retry and self.auto_reconnect:
+                    log.info("Receiver detached. Attempting reconnect.")
+                    await self.reconnect()
+                else:
+                    log.info("Receiver detached. Failed to connect")
+                    error = ConnectError(str(shutdown), shutdown)
+                    raise error
+            except errors.AMQPConnectionError as shutdown:
+                if str(shutdown).startswith("Unable to open authentication session") and self.auto_reconnect:
```

Comment:
Is this really, really the only way to determine that there was an authentication error? Parsing message strings to determine the failure reason has very much a code smell/is not very stable at all...

Reply:
I agree with you. This isn't a good way to use exceptions. I keep this because 
1. this exception was thrown out by uamqp code. I don't want to change this part of uamqp before preview 1. It is not blocking us.
2. we'll refactor code to share a connection among receivers and senders. This part could be impacted. That will be a good timing to change this. 
3. Some methods in Receiver / Sender have very similar "try ... except" code. We'll refactoring this.  Again, not before preview 1.